### ISSUE TITLE
Code style option to prefer default expression over struct default parameterless constructor

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#60592]
**Brief description:**

Offer to simplify `new Struct()` expressions to `default(Struct)` or `default` when `new Struct()` refers to the default parameterless constructor.

The LDM has repeatedly considered treating `new` expressions that refer to the default parameterless constructor as a "trap". We've also heard developers express a preference for the behavior where `new` expressions silently change from meaning "create the default value" to "call the parameterless constructor" when an explicit parameterless constructor is declared.

It's not clear to me if the option should have only "on" and "off" settings, with it offering to change `new()` to `default` one-way when "on". I think it's a little more dubious to offer to change `default` to `new()`, for example, but IDE team is likely to have a more informed opinion on how this should work.

**Languages applicable:**

The option is applicable to C#.

The closest equivalent of `default` in VB is `Nothing`. However, it's not clear if it's similarly idiomatic in VB to replace `New Struct()` with `Nothing` or `CType(Nothing, Struct)`, for example, so perhaps the style option be skipped for VB.

**Code example that the analyzer should report:**

```cs
var s1 = new S(); // offer to replace with `default(S)`
var s2 = new S { x = 1 }; // don't offer to replace

struct S
{
    public int x;
}
```

**Additional information:**

Any more additional information you would like to add.

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

**Comment by TahirAhmadov** [comment]
For:
```cs
struct S
{
    public int x;
}
var s1 = new S(); // doesn't lack of any explicit ctors mean we have an implicit ctor?
// i would think that should be kept
public struct R
{
    public R(int x) => this.x = x;;
    public int x;
}
var r1 = new R(); // on the other hand, I think this should be fixed
```

**Comment by RikkiGibson** [comment]
No, we made a change to avoid synthesizing constructors in structs. So `S` in your example is calling the default parameterless constructor.

**Comment by TahirAhmadov** [comment]
> No, we made a change to avoid synthesizing constructors in structs. So `S` in your example is calling the default parameterless constructor.
Is that in LDM notes anywhere? Or can you please explain why that was decided?

**Comment by RikkiGibson** [comment]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-01-03.md#parameterless-struct-constructors-revisited

**Comment by sharwell** [comment]
This was filed previously as https://github.com/dotnet/runtime/issues/43973. It has also been implemented externally as [SA1129: Do not use default value type constructor](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1129.md).

**Comment by jinujoseph** [comment]
in design meeting to discuss the potential of doing this.

**Comment by dibarbet** [comment]
Design review outcome:

We definitely agree that we should offer a fixer to switch to `default`, however we weren't certain on exactly how we wanted the analyzer to be surfaced and wanted to followup with @RikkiGibson 
1.  Is this actually a quality or a style fix?
2.  Should this be a warning?  Suggestion?
3.  Should this just come as a compiler warning as part of updating to a newer compiler version?  Should we use something like warning waves to surface this warning?
4.  If this is a 'trap' in newer versions of the language, it seems more appropriate to be a compiler warning that we offer a fix for.

@CyrusNajmabadi or @jmarolf please add anything I missed.

**Comment by RikkiGibson** [comment]
> Is this actually a quality or a style fix?

I consider this a style fix. Some users don't consider it categorically better to change `new S()` to `default(S)`, mainly because they *want* the use sites to implicitly change in their meaning if they declare a parameterless constructor later on. cc @jnm2.

> Should this be a warning? Suggestion?

> Should this just come as a compiler warning as part of updating to a newer compiler version? Should we use something like warning waves to surface this warning?

> If this is a 'trap' in newer versions of the language, it seems more appropriate to be a compiler warning that we offer a fix for.

It should be a suggestion for reasons I outlined above. LDM has considered making it a warning in a warning wave but backed off because of the object initializer question (IIRC). I don't consider it a trap or appropriate to include as a compiler or SDK warning for reasons outlined above.

**Comment by CyrusNajmabadi** [comment]
Rikki's reasons seem reasonable to me.  Let's take this as an IDE style option.

**Comment by jmarolf** [comment]
I guess I need to understand @jnm2's case. 

If someone writes this today:
```csharp
ImmutableArray<int> numbers = new();
```

They are actually saying this:
```csharp
ImmutableArray<int> numbers = default;
```

If we were starting from scratch I would very much like this to be normalized: Do not call the constructor that the compiler generates. Just use `default`, they do the same thing and its much more clear that you _intend_ this to happen. 

If, in the future, `ImmutableArray<T>` got a struct constructor (struct 'strutor?) that does the same thing as `Create()` then its clear that you know this and intend to call it when you write this:

```csharp
ImmutableArray<int> numbers = new();
```

Reading the LDM notes: 

> In general, it's a bit unfortunate that C# allows new StructType() syntax. default(T) wasn't in C# 1 (it was introduced with C# 2's generics), so that was the original way to get a default instance of a value type. We think there may be an opportunity to start correcting this with a .NET 7 warning wave, but that will need more design, particularly around object initializers on struct types.

It seems like we are concerned about this case:

```csharp
MyGoodStruct goodStuff = new()
{
	Data = "good-deals.com"
};

struct MyGoodStruct { public string Data; }
```

which I agree should not be converted to this:

```csharp
MyGoodStruct goodStuff = default;
goodStuff.Data = "good-deals.com";

struct MyGoodStruct { public string Data; }
```

So we should not offer any diagnostic (analyzer or otherwise) for this case.

I am fine with it being a style option as it is, essentially, a stylistic choice: the language now has made an old feature redundant and it's no longer necessary to use it. My only other comment is that it's also _extremely confusing_ and I am willing to be a bit more aggressive here and say "we do not recommend calling the default parameterless constructor anymore, instead use `default` unless you are using the object initializer syntax."

I _also_ be fine with this shipping from dotnet/roslyn-analyzers as part of the SDK as a warning wave in .NET 7 but do not want to hold things up on that discussion when there is an urgent need already.

**Comment by TahirAhmadov** [comment]
> If we were starting from scratch I would very much like this to be normalized: Do not call the constructor that the compiler generates. Just use `default`, they do the same thing and its much more clear that you _intend_ this to happen.

Not to sound "pedantic" - I think there is an important distinction. `ImmutableArray` does **not** have a public no-param ctor; the compiler did not generate anything.

> It seems like we are concerned about this case:
> 
> ```cs
> MyGoodStruct goodStuff = new()
> {
> 	Data = "good-deals.com"
> };
> 
> struct MyGoodStruct { public string Data; }
> ```
> 
> which I agree should not be converted to this:
> 
> ```cs
> MyGoodStruct goodStuff = default;
> goodStuff.Data = "good-deals.com";
> 
> struct MyGoodStruct { public string Data; }
> ```

Why not have the compiler synthesize a public no-param ctor in this case?
```cs
struct MyGoodStruct { public string Data; [CompilerGeneratedDefault]public MyGoodStruct() { this = default; } }
var s = new MyGoodStruct  { Data = "abc" }; // this is emitted in IL as: s = default; s.Data = "abc";
```

**Comment by jmarolf** [comment]
> Why not have the compiler synthesize a public no-param ctor in this case?

Performance I believe. 

this
```csharp
var s = new MyGoodStruct { Data = "abc" };

struct MyGoodStruct { public string Data; }
```

translates to this
```csharp
.method private hidebysig static void  '<Main>$'(string[] args) cil managed
{
  .entrypoint
  // Code size       21 (0x15)
  .maxstack  2
  .locals init (valuetype MyGoodStruct V_0)
  IL_0000:  ldloca.s   V_0
  IL_0002:  initobj    MyGoodStruct
  IL_0008:  ldloca.s   V_0
  IL_000a:  ldstr      "abc"
  IL_000f:  stfld      string MyGoodStruct::Data
  IL_0014:  ret
} // end of method Program::'<Main>$'
```

wheras this
```csharp
var s = new MyGoodStruct { Data = "abc" };

struct MyGoodStruct { public string Data; public MyGoodStruct() { this = default; } }
```

becomes this
```csharp
.method private hidebysig static void  '<Main>$'(string[] args) cil managed
{
  .entrypoint
  // Code size       20 (0x14)
  .maxstack  2
  .locals init (valuetype MyGoodStruct V_0)
  IL_0000:  ldloca.s   V_0
  IL_0002:  call       instance void MyGoodStruct::.ctor()
  IL_0007:  ldloca.s   V_0
  IL_0009:  ldstr      "abc"
  IL_000e:  stfld      string MyGoodStruct::Data
  IL_0013:  ret
} // end of method Program::'<Main>$'
```

its the difference between `initobj` and `call` but I do not believe we are willing to silently opt users into slower running code with no way to opt out.

**Comment by jmarolf** [comment]
@TahirAhmadov this would require compiler and spec changes to implement. Please file  start a discussion on https://github.com/dotnet/csharplang about this.

**Comment by RikkiGibson** [comment]
I've become convinced that the LDM should weigh some options here. I'll follow up when I have a proposal.
